automates/program_analysis/CAST2GrFN/visitors/cast_to_agraph_visitor.py

```diff
@@ -0,0 +1,449 @@
+        for n in args:
+            self.G.add_edge(args_node,n)
+        
+        for n in body:
+            self.G.add_edge(body_node,n)
+
+
+        return node_uid
+
+    @visit.register
+    def _(self, node: List):
+        """Visits List nodes. We visit all the elements and add them to
+        this node. This node's UID is returned."""
+        values = []
+        if(node.values != []):
+            values = self.visit_list(node.values)
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid,label="List")
+        for n in values:
+            self.G.add_edge(node_uid,n)
+
+        return node_uid
+
+    @visit.register 
+    def _(self, node: Loop):
+        """Visits Loop nodes. We visit the conditional expression and the
+        body of the loop, and connect them to this node in the graph.
+        This node's UID is returned."""
+        expr = self.visit(node.expr)
+        body = []
+        if(node.body != []):
+            body = self.visit_list(node.body)
+        node_uid = uuid.uuid4()
+        test_uid = uuid.uuid4()
+        body_uid = uuid.uuid4()
+
+        self.G.add_node(node_uid,label="Loop")
+        self.G.add_node(test_uid,label="Test")
+        self.G.add_node(body_uid,label="Body")
+
+        self.G.add_edge(node_uid,test_uid)
+        self.G.add_edge(test_uid,expr)
+        self.G.add_edge(node_uid,body_uid)
+        for n in body:
+            self.G.add_edge(body_uid,n)
+
+        return node_uid
+    
+    @visit.register
+    def _(self, node: ModelBreak):
+        """Visits a ModelBreak (break statment) node. 
+        The node's UID is returned"""
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid, label="Break")
+        return node_uid
+
+    @visit.register
+    def _(self, node: ModelContinue):
+        """Visits a ModelContinue (continue statment) node. 
+        The node's UID is returned"""
+        node_uid = uuid.uuid4()
+        self.G.add_node(node_uid, label="Continue")
+        return node_uid
+
+    @visit.register
+    def _(self, node: ModelIf):
+        """Visits a ModelIf (If statement) node.
+        We visit the condition, and then the body and orelse 
+        attributes if we have any. They're all added to the Graph 
+        accordingly. The node's UID is returned."""
+        expr = self.visit(node.expr)
+        body = []
+        orelse = []
+        if(node.body != []):
+            body = self.visit_list(node.body) 
+        if(node.orelse != None):
+            orelse = self.visit_list(node.orelse)
+
+        node_uid = uuid.uuid4()
+        test_uid = uuid.uuid4()
+        self.G.add_node(node_uid,label="If")
+        self.G.add_node(test_uid,label="Test")
+        self.G.add_edge(node_uid,test_uid)
+        self.G.add_edge(test_uid,expr)
+        
+        body_uid = uuid.uuid4()
+        orelse_uid = uuid.uuid4()
+
+        self.G.add_node(body_uid,label="If Body")
+        self.G.add_node(orelse_uid,label="Else Body")
```

Comment:
This code doesn't necessarily represent if/elifs correctly. For instance:

```
if x == 1:
    ....
elif x == 2:
    ....
elif x == 3:
    ....
```

I think we discussed before this should be a single if node with "cond 1 test" and "cond 1 body", "cond 2 test" and "cond 2 body", etc. until all of the conditions are handled. Then, if there is an else node too, we can have this else body labelled node. The "orelse" field represents the else part of the expression if it is a list of many nodes OR the next condition in the else if if the list has one node that is an if statement in it.

Reply:
I would also prefer to use the method @dpdicken described above.

Reply:
Right, I understand what you mean when it comes to how the visualization should look. I keep getting stuck on how to work the with the If CAST as it is to produce this. I'm going to leave this as it for the moment, and look into this more later.

Reply:
ok, can you add a comment with a TODO about this, because it can only handle IF/IF else statements correctly as written. It would be good to just track that here.